In [5]:
import opendatasets as od

In [7]:
od.download('stackoverflow-developer-survey-2020')

94609408it [00:16, 5897912.68it/s]                               


16384it [00:00, 42778.51it/s]           


8192it [00:00, 24023.62it/s]            


In [ ]:
import pandas as pd

In [ ]:
surver_raw_df = pd.read_csv('stackoverflow-developer-survey-2020\survey_results_public.csv')

In [11]:
surver_raw_df

Respondent                                         MainBranch Hobbyist  \
0               1                     I am a developer by profession      Yes   
1               2                     I am a developer by profession       No   
2               3                        I code primarily as a hobby      Yes   
3               4                     I am a developer by profession      Yes   
4               5  I used to be a developer by profession, but no...      Yes   
...           ...                                                ...      ...   
64456       64858                                                NaN      Yes   
64457       64867                                                NaN      Yes   
64458       64898                                                NaN      Yes   
64459       64925                                                NaN      Yes   
64460       65112                                                NaN      Yes   

        Age Age1stCode CompFreq  CompTotal  ConvertedComp             Country  \
0       NaN         13  Monthly        NaN            NaN             Germany   
1       NaN         19      NaN        NaN            NaN      United Kingdom   
2       NaN         15      NaN        NaN            NaN  Russian Federation   
3      25.0         18      NaN        NaN            NaN             Albania   
4      31.0         16      NaN        NaN            NaN       United States   
...     ...        ...      ...        ...            ...                 ...   
64456   NaN         16      NaN        NaN            NaN       United States   
64457   NaN        NaN      NaN        NaN            NaN             Morocco   
64458   NaN        NaN      NaN        NaN            NaN            Viet Nam   
64459   NaN        NaN      NaN        NaN            NaN              Poland   
64460   NaN        NaN      NaN        NaN            NaN               Spain   

         CurrencyDesc  ...                  SurveyEase           SurveyLength  \
0       European Euro  ...  Neither easy nor difficult  Appropriate in length   
1      Pound sterling  ...                         NaN                    NaN   
2                 NaN  ...  Neither easy nor difficult  Appropriate in length   
3        Albanian lek  ...                         NaN                    NaN   
4                 NaN  ...                        Easy              Too short   
...               ...  ...                         ...                    ...   
64456             NaN  ...                         NaN                    NaN   
64457             NaN  ...                         NaN                    NaN   
64458             NaN  ...                         NaN                    NaN   
64459             NaN  ...                         NaN                    NaN   
64460             NaN  ...                         NaN                    NaN   

      Trans                                     UndergradMajor  \
0        No  Computer science, computer engineering, or sof...   
1       NaN  Computer science, computer engineering, or sof...   
2       NaN                                                NaN   
3        No  Computer science, computer engineering, or sof...   
4        No  Computer science, computer engineering, or sof...   
...     ...                                                ...   
64456   NaN  Computer science, computer engineering, or sof...   
64457   NaN                                                NaN   
64458   NaN                                                NaN   
64459   NaN                                                NaN   
64460   NaN  Computer science, computer engineering, or sof...   

            WebframeDesireNextYear                      WebframeWorkedWith  \
0                     ASP.NET Core                    ASP.NET;ASP.NET Core   
1                              NaN                                     NaN   
2                              NaN                           

What are the columns present in the dataframe.

In [12]:
surver_raw_df.columns

Index(['Respondent', 'MainBranch', 'Hobbyist', 'Age', 'Age1stCode', 'CompFreq',
       'CompTotal', 'ConvertedComp', 'Country', 'CurrencyDesc',
       'CurrencySymbol', 'DatabaseDesireNextYear', 'DatabaseWorkedWith',
       'DevType', 'EdLevel', 'Employment', 'Ethnicity', 'Gender', 'JobFactors',
       'JobSat', 'JobSeek', 'LanguageDesireNextYear', 'LanguageWorkedWith',
       'MiscTechDesireNextYear', 'MiscTechWorkedWith',
       'NEWCollabToolsDesireNextYear', 'NEWCollabToolsWorkedWith', 'NEWDevOps',
       'NEWDevOpsImpt', 'NEWEdImpt', 'NEWJobHunt', 'NEWJobHuntResearch',
       'NEWLearn', 'NEWOffTopic', 'NEWOnboardGood', 'NEWOtherComms',
       'NEWOvertime', 'NEWPurchaseResearch', 'NEWPurpleLink', 'NEWSOSites',
       'NEWStuck', 'OpSys', 'OrgSize', 'PlatformDesireNextYear',
       'PlatformWorkedWith', 'PurchaseWhat', 'Sexuality', 'SOAccount',
       'SOComm', 'SOPartFreq', 'SOVisitFreq', 'SurveyEase', 'SurveyLength',
       'Trans', 'UndergradMajor', 'WebframeDesireNextYear',
  

In [13]:
surver_raw_df.shape

(64461, 61)

In [ ]:
schema_raw = pd.read_csv('stackoverflow-developer-survey-2020\survey_results_schema.csv', index_col = 'Column').QuestionText

In [19]:
schema_raw.shape

(61,)

Now we can we use schema_raw to retrive the full question.

In [20]:
schema_raw['YearsCodePro']

'NOT including education, how many years have you coded professionally (as a part of your work)?'

## Data Preparation & Cleaning

While the survey responses contain a wealth of information, we'll limit our analysis to the following areas:

- Demographics of the survey respondents and the global programming community
- Distribution of programming skills, experience, and preferences
- Employment-related information, preferences, and opinions

Let's select a subset of columns with the relevant data for our analysis.

In [21]:
selected_columns = [
    # Demographics
    'Country',
    'Age',
    'Gender',
    'EdLevel',
    'UndergradMajor',
    # Programming experience
    'Hobbyist',
    'Age1stCode',
    'YearsCode',
    'YearsCodePro',
    'LanguageWorkedWith',
    'LanguageDesireNextYear',
    'NEWLearn',
    'NEWStuck',
    # Employment
    'Employment',
    'DevType',
    'WorkWeekHrs',
    'JobSat',
    'JobFactors',
    'NEWOvertime',
    'NEWEdImpt'
]

In [22]:
survey_df = surver_raw_df[selected_columns].copy()

In [23]:
survey_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64461 entries, 0 to 64460
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Country                 64072 non-null  object 
 1   Age                     45446 non-null  float64
 2   Gender                  50557 non-null  object 
 3   EdLevel                 57431 non-null  object 
 4   UndergradMajor          50995 non-null  object 
 5   Hobbyist                64416 non-null  object 
 6   Age1stCode              57900 non-null  object 
 7   YearsCode               57684 non-null  object 
 8   YearsCodePro            46349 non-null  object 
 9   LanguageWorkedWith      57378 non-null  object 
 10  LanguageDesireNextYear  54113 non-null  object 
 11  NEWLearn                56156 non-null  object 
 12  NEWStuck                54983 non-null  object 
 13  Employment              63854 non-null  object 
 14  DevType                 49370 non-null

In [25]:
survey_df['Age1stCode'] = pd.to_numeric(survey_df.Age1stCode, errors='coerce')
survey_df['YearsCode'] = pd.to_numeric(survey_df.YearsCode, errors='coerce')
survey_df['YearsCodePro'] = pd.to_numeric(survey_df.YearsCodePro, errors='coerce')

In [26]:
survey_df.describe()

Age    Age1stCode     YearsCode  YearsCodePro   WorkWeekHrs
count  45446.000000  57473.000000  56784.000000  44133.000000  41151.000000
mean      30.834111     15.476572     12.782051      8.869667     40.782174
std        9.585392      5.114081      9.490657      7.759961     17.816383
min        1.000000      5.000000      1.000000      1.000000      1.000000
25%       24.000000     12.000000      6.000000      3.000000     40.000000
50%       29.000000     15.000000     10.000000      6.000000     40.000000
75%       35.000000     18.000000     17.000000     12.000000     44.000000
max      279.000000     85.000000     50.000000     50.000000    475.000000

Here we see that the minimum age is 1 and maximum age is 279. These are errors in  the surveys.  A simple fix would be to ignore the rows where the age is higher than 100 years or lower than 10 years as invalid survey responses. 

In [30]:
survey_df.drop(survey_df[survey_df.Age < 10].index, inplace=True)
survey_df.drop(survey_df[survey_df.Age > 100].index, inplace=True)

The same holds for `WorkWeekHrs`. Let's ignore entries where the value for the column is higher than 140 hours. (~20 hours per day).

In [31]:
survey_df.drop(survey_df[survey_df.WorkWeekHrs > 140].index, inplace=True)

The gender column also allows for picking multiple options. We'll remove values containing more than one option to simplify our analysis.

In [32]:
survey_df.Gender.value_counts()

Gender
Man                                                            45895
Woman                                                           3835
Non-binary, genderqueer, or gender non-conforming                385
Man;Non-binary, genderqueer, or gender non-conforming            121
Woman;Non-binary, genderqueer, or gender non-conforming           92
Woman;Man                                                         73
Woman;Man;Non-binary, genderqueer, or gender non-conforming       25
Name: count, dtype: int64

In [33]:
import numpy as np 

In [34]:
survey_df.where(~(survey_df.Gender.str.contains(';', na=False)), np.nan, inplace=True)

Now that we have cleared the dataset for our analysis, let's look at the dataset sample for some understanding.

In [35]:
survey_df.sample(10)

Country   Age Gender  \
2317           Brazil  24.0    Man   
12        Netherlands  53.0    Man   
61350   United States   NaN    Man   
3052    United States  54.0    Man   
51440       Sri Lanka   NaN    NaN   
30421   United States   NaN    Man   
18924  United Kingdom   NaN    Man   
58864           India   NaN    NaN   
5558            Japan   NaN    NaN   
5404           Canada  27.0    Man   

                                                 EdLevel  \
2317                                                 NaN   
12     Secondary school (e.g. American high school, G...   
61350       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3052        Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
51440                                                NaN   
30421                                                NaN   
18924  Secondary school (e.g. American high school, G...   
58864  Secondary school (e.g. American high school, G...   
5558        Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
5404        Bachelor’s degree (B.A., B.S., B.Eng., etc.)   

                                          UndergradMajor Hobbyist  Age1stCode  \
2317   Information systems, information technology, o...      Yes        15.0   
12                                                   NaN      Yes        14.0   
61350  Computer science, computer engineering, or sof...      Yes         8.0   
3052                           Mathematics or statistics       No        16.0   
51440                                                NaN      Yes         NaN   
30421  Computer science, computer engineering, or sof...      Yes        16.0   
18924                                                NaN      Yes         7.0   
58864                                                NaN      Yes        17.0   
5558   Computer science, computer engineering, or sof...      Yes        13.0   
5404   Computer science, computer engineering, or sof...      Yes        17.0   

       YearsCode  YearsCodePro  \
2317        10.0           1.0   
12          35.0          20.0   
61350       33.0          22.0   
3052        39.0          35.0   
51440        NaN           NaN   
30421        3.0           NaN   
18924       10.0           NaN   
58864        3.0           NaN   
5558        20.0          20.0   
5404        10.0           4.0   

                                      LanguageWorkedWith  \
2317               C#;HTML/CSS;JavaScript;SQL;TypeScript   
12                                   C;JavaScript;Python   
61350  Bash/Shell/PowerShell;C;C#;C++;Go;HTML/CSS;Jav...   
3052                                   HTML/CSS;Java;SQL   
51440                                                NaN   
30421  Assembly;Bash/Shell/PowerShell;C;C#;C++;Java;P...   
18924                                 C#;HTML/CSS;Python   
58864                                                NaN   
5558                          HTML/CSS;JavaScript;Python   
5404                          Bash/Shell/PowerShell;Java   

                                  LanguageDesireNextYear  \
2317   Bash/Shell/PowerShell;C#;Go;HTML/CSS;Java;Java...   
12                                                Python   
61350                     C++;Dart;Go;Kotlin;Python;Rust   
3052                                   HTML/CSS;Java;SQL   
51440                                                NaN   
30421                                  Objective-C;Swift   
18924                                    Java;JavaScript   
58864                                                NaN   
5558                                                 NaN   
5404                          Bash/Shell/PowerShell;Java   

                   NEWLearn  \
2317       Every few months   
12     Once every few years   
61350  Once every few years   
3052   Once every few years   
51440                   NaN   
30421           Once a year   
18924           Once a year   
58864                   NaN   
5558          Once a decade   
5404       Every few months   

   